packages/salesforcedx-vscode-core/src/commands/baseDeployRetrieve.ts

```diff
@@ -0,0 +1,390 @@
+        'metadataCount',
+        this.createComponentCount(components)
+      );
+
+      result = await this.doOperation(components);
+
+      const status = this.getStatus(result);
+
+      return (
+        status === RequestStatus.Succeeded ||
+        status === RequestStatus.SucceededPartial
+      );
+    } finally {
+      await this.postOperation(result);
+    }
+  }
+
+  protected getRelativeProjectPath(fsPath: string = '', packageDirs: string[]) {
+    let packageDirIndex;
+    for (let packageDir of packageDirs) {
+      if (!packageDir.startsWith(sep)) {
+        packageDir = sep + packageDir;
+      }
+      if (!packageDir.endsWith(sep)) {
+        packageDir = packageDir + sep;
+      }
+      packageDirIndex = fsPath.indexOf(packageDir);
+      if (packageDirIndex !== -1) {
+        packageDirIndex += 1;
+        break;
+      }
+    }
+    return packageDirIndex !== -1 ? fsPath.slice(packageDirIndex) : fsPath;
+  }
+
+  private createComponentCount(
+    components: Iterable<MetadataComponent>
+  ): string {
+    const quantities: { [type: string]: number } = {};
+    for (const component of components) {
+      const { name: typeName } = component.type;
+      const typeCount = quantities[typeName];
+      quantities[typeName] = typeCount ? typeCount + 1 : 1;
+    }
+    return JSON.stringify(
+      Object.keys(quantities).map(type => ({
+        type,
+        quantity: quantities[type]
+      }))
+    );
+  }
+
+  private getStatus(
+    result: DeployRetrieveResult | undefined
+  ): RequestStatus | undefined {
+    return result && 'response' in result
+      ? result.response.status
+      : result?.status;
+  }
+
+  protected abstract getComponents(
+    response: ContinueResponse<T>
+  ): Promise<ComponentSet>;
+  protected abstract doOperation(
+    components: ComponentSet
+  ): Promise<DeployRetrieveResult | undefined>;
+  protected abstract postOperation(
+    result: DeployRetrieveResult | undefined
+  ): Promise<void>;
+}
+
+export abstract class DeployCommand<T> extends DeployRetrieveExecutor<T> {
+  protected async doOperation(
+    components: ComponentSet
+  ): Promise<DeployResult | undefined> {
+    return components
+      .deploy({
+        usernameOrConnection: await workspaceContext.getConnection()
+      })
+      .start();
+  }
+
+  protected async postOperation(
+    result: DeployResult | undefined
+  ): Promise<void> {
+    try {
+      if (result) {
+        BaseDeployExecutor.errorCollection.clear();
+
+        const relativePackageDirs = await SfdxPackageDirectories.getPackageDirectoryPaths();
+        const output = this.createOutput(result, relativePackageDirs);
+        channelService.appendLine(output);
+
+        const success = result.response.status === RequestStatus.Succeeded;
+
+        if (!success) {
+          handleDeployDiagnostics(result, BaseDeployExecutor.errorCollection);
+        }
+      }
+    } finally {
+      await DeployQueue.get().unlock();
+    }
```

Comment:
All deploy commands need to:
- output results the same way
- handle diagnostics the same way
- unlock the deploy queue when they are finished